### ISSUE TITLE
Allow disabling checks from comments

### ISSUE THREAD

**Issue by HackAttack** [CodeNarc/CodeNarc#156]
It would be nice to have the ability to selectively disable checks via comments (e.g. Checkstyle, ESLint).

**Comment by chrismair** [comment]
Is using the @SuppressWarnings annotation not sufficient?
http://codenarc.sourceforge.net/codenarc-configuring-rules.html#Suppressing_A_Rule_From_Within_Source_Code

**Comment by HackAttack** [comment]
The limitation with that is it’s coarse-grained compared to suppressing a check for an individual line or handful of lines. Ideally when checks are ignored it’s scoped to as small a thing as possible, but the annotation can only be applied to things like classes and methods (not, for example, local variables).

The actual use case that spurred this feature request was suppressing the line length check in a particular comment that contained a URL. I know I could have configured the line length check to ignore lines containing “http,” but I would want to make disabling checks surgical if possible.

**Comment by w4-jungyeol** [comment]
The same issue. For example, it's not easy to disable LineLength checker in this code.
/**
 * Blah Blah...
 * https://www.example.com/path/to/some/long/urls/long.html?query1=a&query2=b&query3=c&query4=d
 */

**Comment by chrismair** [comment]
Adding a comment-based rule-enablement facility to CodeNarc would have pretty broad consequences, if it was generic and not rule-specific.  

Given that both examples so far are for *LineLength*, would it work to add whatever tag or string you'd want to use to the end of those long lines and then configure the rule's *ignoreLineRegex* property to ignore those lines?
e.g., using "LineLength" as the tag:

```
/**
 * Blah Blah...
 * https://www.example....  // LineLength
 */
```
and then:
```
LineLength {
    ignoreLineRegex = /LineLength/
}

```

**Comment by HackAttack** [comment]
That would certainly work as a stopgap, and I agree that this is a hefty feature request.

**Comment by jlstrater** [comment]
Since this issue is rather old and there aren't any new comments, is this still needed or should it be closed?

**Comment by decoded4620** [comment]
I think this feature is essential. In my unit tests, specifically i need to test certain things and test resources (because they are dummy) often violate stupid rules, (EmptyClass, UnneccessaryPublicModifier for example). 

// CODENARC: ON ... // CODENARC: OFF 

Or something similar is imperative as a feature (I would think) for these types of nitty cases, which is why other static analysis tools provide it. We should be on Par

**Comment by maharvey** [comment]
Agreed, this feature is essential. We have policy to allow no Codenarc violations, but sometimes Codenarc insists on things that make the code un-buildable. For example we have a "best practice" to use @CompileStatic in groovy, but codenarc insists on stylistic features that break when static compilation is used.

**Comment by chrismair** [comment]
So that I better understand the requirements/issue ... Are these cases where you are unable to disable rules [based on class or filenames](http://codenarc.sourceforge.net/codenarc-configuring-rules.html#Standard_Properties_for_Configuring_Rules)? Do you have control over the ruleset used and/or the "codenarc.properties" file? Are these cases where you can use @SuppressWarnings at the class level? And is this a particular set of troublesome rules, or is it truly a global requirement on all rules?

**Comment by sfrenk** [comment]
Just wanted to chime in and add a vote for this feature. All of the linters we currently use (Pylint, shellcheck and LintR) allow the disabling of specific rules at specific lines using comments and we use this functionality regularly. It is generally preferable to be able to disable rules at the line level rather than the class or file level as this functionality is generally useful for very specific cases (eg. a long string which is much more readable if the line is kept at over 80 characters).

**Comment by unitysipu** [comment]
This feature is pretty critical to allow targeted / transparent ignores of individual issues in code.

npm-groovy-lint wrapper has a pretty elegant solution for this problem (and also a VSCode plugin that makes annotating code / ignoring things super easy), I suggest checking it out for inspiration. https://github.com/nvuillam/npm-groovy-lint#disabling-rules-in-source

**Comment by nvuillam** [comment]
@unitysipu  thanks for the compliment ;)

But to render to Caesar the things that are Caesar's , I took a big inspiration from [eslint disable pattern](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments), just replacing eslint by groovylint 😄

**Comment by chrismair** [comment]
I agree that the npm-groovy-lint capabilities are pretty impressive and inspiring (nice work Nicolas!). I have to admit I was thinking about a much simpler capability, more like the PMD (//NOPMD) or Sonar (//NOSONAR) affecting only the current line. But a more sophisticated solution does seem feasible and deserves consideration; I'll be taking a deeper look at that. Thanks.
